### PR TITLE
API-1124 add bulk transfer

### PR DESCRIPTION
**Pull request by wshino**
add schema for bulk transfer

### CODE REVIEW

File: subgraph.template.yaml [hunk 1]
```diff
@@ -152,3 +152,25 @@ dataSources:
           handler: handleCreateSRRWithProof
         - event: MigrateBatch(indexed bytes32,bytes32[],address,uint256,uint256)
           handler: handleMigrateBatch
+  - kind: ethereum/contract
```
Comment (noyonthe1) [inline subgraph.template.yaml:155]: So, we will have a new contract entry in that case I think deployment.json should have contract entry as well.

File: deployments.json [hunk 2]
```diff
@@ -38,7 +44,10 @@
       "startBlock": 13286271
     },
     "BulkIssue": {
-      "startBlock": 13286271
+      "startBlock": 11907787
```
Comment (chatch) [inline deployments.json:47]: better to make these start blocks the same as the others. ideally BulkTransfer will get set to the block it is created at, when that happens ... but for now this placeholder can be the same height as previously deployed contracts.

Comment (chatch) [reply deployments.json:47]: for BulkIssue and BulkTransfer ...

File: deployments.json [hunk 3]
```diff
@@ -25,6 +28,9 @@
     },
     "BulkIssue": {
       "startBlock": 12378987
+    },
+    "BulkTransfer": {
+      "startBlock": 11907787
```
Comment (chatch) [inline deployments.json:33]: here too (see polygon section comment) this should be at least as height as the other contracts but ideally set to the height the contract is deployed at.

File: schema.graphql [hunk 4]
```diff
@@ -125,6 +125,16 @@ type BulkIssue @entity {
   updatedAt: BigInt! # block.timestamp when entity last updated
 }
 
+type BulkTransfer @entity {
+  id: ID! # merkleRoot.toHexString()
+  merkleRoot: Bytes! # batch merkle root
+  txs: [Bytes!] # Hashes of ApproveSRRs processed
```
Comment (chatch) [inline schema.graphql:131]: `txs` doesn't really reflect what is stored here i think. how about srrApproveHashes as this is what is stored here. the event param name is also `srrApproveHash`

File: src/bulk-transfer-mapping.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  batch = new BulkTransfer(merkleRoot);
+  batch.txs = [];
+  batch.merkleRoot = event.params.merkleRoot;
+  batch.sender = event.params.sender;
+
+  batch.createdAt = batch.updatedAt = eventUTCMillis(event);
+  batch.save();
+}
+
+export function handleCreateSRRWithProof(event: ApproveSRRByCommitmentWithProofEvent): void {
```
Comment (chatch) [inline src/bulk-transfer-mapping.ts:32]: createSRR in BulkTransfer?  this is handleApprove... right?

Comment (chatch) [reply src/bulk-transfer-mapping.ts:32]: update name used in the logging below too. logInvocation and log.error ...

File: src/bulk-transfer-mapping.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  // (see https://thegraph.com/docs/assemblyscript-api#api-reference):
+  let txs = batch.txs;
+  txs.push(event.params.srrApproveHash);
+  batch.txs = txs;
+  batch.tokenId = event.params.tokenId.toString();
+  batch.updatedAt = eventUTCMillis(event);
+  batch.save();
+}
+
+export function handleMigrateBatch(event: MigrateBatchEvent): void {
```
Comment (chatch) [inline src/bulk-transfer-mapping.ts:56]: this is not needed. migration is already done so there won't be a BulkTransfer migrate.

File: subgraph.template.yaml [hunk 7]
```diff
@@ -152,3 +152,25 @@ dataSources:
...
+        - SRR
+      abis:
+        - name: BulkTransfer
+          file: ./abis/BulkTransfer.json
+      file: ./src/bulk-transfer-mapping.ts
+      eventHandlers:
+        - event: BatchPrepared(indexed bytes32,indexed address)
+          handler: handleBatchPrepared
+        - event: ApproveSRRByCommitmentWithProof(indexed bytes32,indexed uint256,bytes32)
+          handler: handleCreateSRRWithProof
```
Comment (chatch) [inline subgraph.template.yaml:176]: here too - rename to approve ..

File: subgraph.template.yaml (hunk 1 above)
Comment (wshino) [reply subgraph.template.yaml:155]: @noyonthe1 thanks, I added the BulkTransfer into deployments.json
https://github.com/startbahn/startrail-subgraph/pull/64/files#diff-0459c08daaef16ee190d4991e600a598e34e76a05bd2d122c38030ad040ebea1R15

File: src/bulk-transfer-mapping.ts [hunk 8]
```diff
@@ -0,0 +1,71 @@
+import { log } from '@graphprotocol/graph-ts'
+
+import {
+  BatchPrepared as BatchPreparedEvent,
+  ApproveSRRByCommitmentWithProof as ApproveSRRByCommitmentWithProofEvent,
+  MigrateBatch as MigrateBatchEvent,
```
Comment (chatch) [inline src/bulk-transfer-mapping.ts:6]: remove this event too - it's not used.

File: subgraph.template.yaml (hunk 1 above)
Comment (noyonthe1) [reply subgraph.template.yaml:155]: oh I see, sorry I missed that part then.

File: src/bulk-transfer-mapping.ts (hunk 8 above)
Comment (chatch) [reply src/bulk-transfer-mapping.ts:6]: oh it's on the contract too - let's remove it from BulkTransfer,sol too but don't need to redeploy until after the testing i think. it's just unused code.

**Review by chatch — CHANGES_REQUESTED** [review]

File: deployments.json (hunk 2 above)
Comment (wshino) [reply deployments.json]: ah, I set old block, fixed it.

File: schema.graphql (hunk 4 above)
Comment (wshino) [reply schema.graphql:131]: I agree with you, fixed it.

File: src/bulk-transfer-mapping.ts (hunk 5 above)
Comment (wshino) [reply src/bulk-transfer-mapping.ts:32]: Sorry, it. is a copy from bulk, fixed it.

File: src/bulk-transfer-mapping.ts (hunk 6 above)
Comment (wshino) [reply src/bulk-transfer-mapping.ts:56]: I think so, removed.

File: src/bulk-transfer-mapping.ts (hunk 8 above)
Comment (wshino) [reply src/bulk-transfer-mapping.ts:6]: Okay, I removed it later from contract.

**Review by chatch — APPROVED** [review]
great stuff - approved after changes

**Pull request merged**